src/preprocessing/passes/symmetry_detect.cpp

```diff
@@ -21,22 +21,483 @@ using namespace std;
+  unsigned base_occurs_req = d_base_vars.size();
+  do
+  {
+    Trace("sym-dt-debug") << "   base variables must occur " << base_occurs_req
+                          << " times." << std::endl;
+    // check if all the base_vars occur at least the required number of
+    // times
+    bool var_ok = true;
+    for (const Node& v : d_base_vars)
+    {
+      if (d_occurs_count[v] < base_occurs_req)
+      {
+        Trace("sym-dt-debug") << "...failed variable " << v << std::endl;
+        var_ok = false;
+        break;
+      }
+    }
+    if (!var_ok)
+    {
+      // cannot merge due to a base variable
+      break;
+    }
+    // try to find a new variable to merge
+    Trace("sym-dt-debug") << "   must find " << d_num_new_indices_needed
+                          << " new indices to merge." << std::endl;
+    std::vector<unsigned> new_indices;
+    Node merge_var;
+    d_merge_var_tried.clear();
+    if (mergeNewVar(0, new_indices, merge_var, 0, partitions, active_indices))
+    {
+      Trace("sym-dt-debug") << "   ...merged: " << merge_var << std::endl;
+      Assert(!merge_var.isNull());
+      merged = true;
+      success = true;
+      // update the number of new indicies needed
+      if (base_choose > 0)
+      {
+        d_num_new_indices_needed +=
+            nChoosek(d_base_vars.size(), base_choose - 1);
+        // base_occurs_req =   TODO
+      }
+    }
+    else
+    {
+      Trace("sym-dt-debug") << "   ...failed to merge" << std::endl;
+      success = false;
+    }
+  } while (success);
+  return merged;
+}
+
+bool PartitionMerger::mergeNewVar(unsigned curr_index,
+                                  std::vector<unsigned>& new_indices,
+                                  Node& merge_var,
+                                  unsigned num_merge_var_max,
+                                  std::vector<Partition>& partitions,
+                                  std::unordered_set<unsigned>& active_indices)
+{
+  Assert(new_indices.size() < d_num_new_indices_needed);
+  if (curr_index == d_indices.size())
+  {
+    return false;
+  }
+  Trace("sym-dt-debug2") << "merge " << curr_index << " / " << d_indices.size()
+                         << std::endl;
+  // try to include this index
+  unsigned index = d_indices[curr_index];
+
+  // if not already included
+  if (d_base_indices.find(index) == d_base_indices.end())
+  {
+    Assert(active_indices.find(index) != active_indices.end());
+    // check whether it can merge
+    Partition& p = partitions[index];
+    Assert(p.d_subvar_to_vars.size() == 1);
+    std::vector<Node>& svs = p.d_subvar_to_vars.begin()->second;
+    bool include_success = true;
+    Node curr_merge_var;
+    for (const Node& v : svs)
+    {
+      if (d_base_vars.find(v) == d_base_vars.end() && v != merge_var)
+      {
+        if (merge_var.isNull() && curr_merge_var.isNull())
+        {
+          curr_merge_var = v;
+        }
+        else
+        {
+          // cannot include
```

Comment:
Here, we cannot include, is it because the current partition p have 2 or more new variables in its partition?

Reply:
Yes, recall in this method, we are trying to find a set of partitions that suffice to add exactly one new variable to the region. This case holds when adding the current region would cause 2+ variables to be added, hence, we cannot add this region.